src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs

```diff
@@ -231,10 +229,13 @@ public static async ValueTask WriteResponseMessage(HttpResponse response, Encodi
                     }
                 }
 
-                if (serverCallContext.DescriptorInfo.BodyFieldDescriptors != null)
+                if (serverCallContext.DescriptorInfo.BodyFieldDescriptor != null)
                 {
                     requestMessage = (IMessage)Activator.CreateInstance<TRequest>();
-                    ServiceDescriptorHelpers.RecursiveSetValue(requestMessage, serverCallContext.DescriptorInfo.BodyFieldDescriptors, bodyContent); // TODO - check nullability
+
+                    // The spec says that request body must be on the top-level message.
+                    // Recursive request body isn't supported.
+                    ServiceDescriptorHelpers.SetValue(requestMessage, serverCallContext.DescriptorInfo.BodyFieldDescriptor, bodyContent);
```

Comment:
Do we need to check that `BodyFieldDescriptor` isn't null before calling out to `SetValue`?

Reply:
It's checked for null a few lines before this.